File: longest_zigzag_path_in_a_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int longestZigZag(TreeNode* root) {
        if (!root) {
            return 0;
        }

        int max_length = 0;

        dfs(root, true, 0, max_length);
        return max_length;
    }

private:
    void dfs(TreeNode* node, bool is_left, int length, int& max_length) {
        if (!node) {
            return;
        }

        max_length = std::max(max_length, length);

        if (is_left) {
            dfs(node->left, false, length + 1, max_length);
            dfs(node->right, true, 1, max_length);
        } else {
            dfs(node->right, true, length + 1, max_length);
            dfs(node->left, false, 1, max_length);
        }
    }    
};

File: longest_zigzag_path_in_a_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:

        if not root:
            return 0
        
        max_length = 0

        def dfs(node, is_left, length):
            nonlocal max_length

            if not node:
                return
            
            max_length = max(max_length, length)

            if is_left:
                dfs(node.left, False, length + 1)
                dfs(node.right, True, 1)
            else:
                dfs(node.right, True, length + 1)
                dfs(node.left, False, 1)                

        dfs(root, True, 0)
        return max_length
